Multiple orientation support

maartenslaa opened the pull request:
I want to share what I'm working on and hope this is the best way to do it. I made changes from the master branche v1.4.1 to support multiple orientations for both Android and iOS. This is my first pull requests ever so feel free to educate me if I should do something different next time. I grouped all changes in a single commit "Multiple orientation support" but the other commits show as well, is there a way to only show this commit in the pull request? Or does it require me to create a new branche?

The head tracker needs to know the start orientation when the app is opened. I'm not sure why but the ekf_to_head_tracker needs to be different for each orientation when initialised. It is therefore set in the constructor. Depending on the current orientation the sensor_to_display is set in GetPose().

I've added some functionality to screen_params:
For Android I added a ScreenOrientation object with an OrientationEventListener and only call CurrentOrientation when the orientation has changed because defaultDisplay.getRotation() takes a long time to execute.

For iOS I added an OrientationHelper object which gets notified when the orientation changes. (I now realise I should have named it ScreenOrientation as well like the Android one.)

The above objects keep track of the current orientation and return their orientation through a call to getScreenOrientation in screen_params.h.

The head tracker now supports landscape left, portrait and landscape right. Rendering supports landscape left and landscape right, portrait may require changes to the render texture dimensions but yeah, who needs stereoscopic portrait mode anyway.

Disclaimer: I'm no expert on multi thread safety so keep an eye out for any issues that may arise from that.

maartenslaa commented:
So I should have done a test build of the commit, there are still some small errors:

[In the constructor a ')' is missing after GyroscopeData()](https://github.com/Aryzon/cardboard/blob/43afbe0d87dd5311aa08cb32e0127420893fd6a1/sdk/head_tracker.cc#L31)
[Rotation ekf_to_head_tracker; is missing here](https://github.com/Aryzon/cardboard/blob/5d43e37e28ab92e1ddf8c3c6d7d2bd962e85fab2/sdk/head_tracker.h#L91)
[private static Display defaultDisplay; is missing here](https://github.com/Aryzon/cardboard/blob/5d43e37e28ab92e1ddf8c3c6d7d2bd962e85fab2/sdk/screen_params/android/java/com/google/cardboard/sdk/screenparams/ScreenParamsUtils.java#L115)

agalbachicar commented on sdk/screen_params/android/java/com/google/cardboard/sdk/screenparams/ScreenParamsUtils.java:
```diff
@@ -68,4 +75,81 @@ public static ScreenPixelDensity getScreenPixelDensity(Context context) {
+      }
+
+      int orientation = context.getResources().getConfiguration().orientation;
+      // This call takes a long time so we don't call this every frame, only when rotation changes
```
Can you move this docstring to the method docstring?

agalbachicar commented on sdk/screen_params/ios/screen_params.mm:
```diff
@@ -18,6 +18,74 @@
 #import <sys/utsname.h>
 
+// Aryzon multiple orientations
+@interface OrientationHelper : NSObject {
```
Would you call it `ScreenOrientationHelper`?

agalbachicar reviewed (changes requested):
Thanks a lot for your contribution. I've done a first pass to the code. PTAL.

Let me reply in between lines to some of your comments in the PR description as well.

---------

> I want to share what I'm working on and hope this is the best way to do it. I made changes from the master branche v1.4.1 to support multiple orientations for both Android and iOS. 

Great! That's the right place to start working on.

> This is my first pull requests ever so feel free to educate me if I should do something different next time.

No problem, we will help you go through it. Note that the scope might increase or decrease a bit of your PRs due to missing functionality coverage.


> I grouped all changes in a single commit "Multiple orientation support" but the other commits show as well, is there a way to only show this commit in the pull request? Or does it require me to create a new branche?

It depends on what you want to achieve. If your changes depend one on another, I suggest you create a hierarchy of branches like the following:

<pre>
master <-- branch_1 <-- branch_2 <-- branch_3 ...
</pre>

> The head tracker needs to know the start orientation when the app is opened. I'm not sure why but the ekf_to_head_tracker needs to be different for each orientation when initialised. It is therefore set in the constructor. Depending on the current orientation the sensor_to_display is set in GetPose().

This PR is missing the interaction with Unity. It has it's own user-defined configuration that will affect the applications flow. I recommend making the transformation outside the `HeadTracker::GetPose()`method for the moment. Until that is solved.


> I've added some functionality to screen_params:
For Android I added a ScreenOrientation object with an OrientationEventListener and only call CurrentOrientation when the orientation has changed because defaultDisplay.getRotation() takes a long time to execute.

I left some comments there.

> For iOS I added an OrientationHelper object which gets notified when the orientation changes. (I now realise I should have named it ScreenOrientation as well like the Android one.)

I left a comment there as well.

> The above objects keep track of the current orientation and return their orientation through a call to getScreenOrientation in screen_params.h.

That's the right way to implement it.

> The head tracker now supports landscape left, portrait and landscape right. Rendering supports landscape left and landscape right, portrait may require changes to the render texture dimensions but yeah, who needs stereoscopic portrait mode anyway.

We noted that. Let me share some results of our experimentation with you.

- We needed to change [AndroidManifest.xml#24](https://github.com/googlevr/cardboard/blob/master/hellocardboard-android/src/main/AndroidManifest.xml#L24) from `landscape` to `userLandscape`. Have you done that?
- On portrait, as you mentioned, the texture dimensions are wrong. The application will be in charge of setting a new set of CardboardMesh descriptions when the screen changes. We would ask you to provide at the `cardboard.h` a new method to get the screen orientation enum. That way, it will be the application the responsible of re-configuring the SDK to have a proper visualisation. It might even skip the portrait mode and just toggle between landscape left and right. 
- We noticed that when you move from landscape left to portrait and then from portrait to landscape right it works but when you move from landscape right to left directly the scene is upside down. Would you please check that?

> Disclaimer: I'm no expert on multi thread safety so keep an eye out for any issues that may arise from that.

I'll do another complete review specifically for that

maartenslaa commented:
Thanks for going over the code! It really helps me learn the correct way of writing. I will be implementing the changes you suggest this week.

maartenslaa commented:
I think I have implemented all your suggestions. Thanks again for all you comments, they are clear and good to follow. Some answers / comments on your notes:

- We needed to change AndroidManifest.xml#24 from landscape to userLandscape. Have you done that?

I deleted the line entirely to also use portrait mode.

- On portrait, as you mentioned, the texture dimensions are wrong. The application will be in charge of setting a new set of CardboardMesh descriptions when the screen changes. We would ask you to provide at the cardboard.h a new method to get the screen orientation enum. That way, it will be the application the responsible of re-configuring the SDK to have a proper visualisation. It might even skip the portrait mode and just toggle between landscape left and right.

I may have been mistaken, I think on Android the dimensions are correct. On iOS it was not, I fixed that by making some changes in the HelloCardboardViewController. Have a look at [this](https://github.com/googlevr/cardboard/pull/208/commits/7c919a642ce780a3e5fb1a0bf861dc0b9d41743e) commit for the changes.

- We noticed that when you move from landscape left to portrait and then from portrait to landscape right it works but when you move from landscape right to left directly the scene is upside down. Would you please check that?

Yes I hadn't noticed! I made some changes to fix it, though I am a little bit unhappy with the result because I had to use rotation angles for which I could not find constants. I do think I works quite well now.

I also added a method to get the screen orientation through cardboard.h like you said. I don't think it is necessary to use the enum however, on Android the current implementation seems to work correctly. On iOS the changes to HelloCardboardViewController (see above) seem to fix it as well.

You mentioned to use another method instead of GetPose on the headtracker because this is missing the Unity interaction. I have tested with Unity and to my knowledge this did not impact the Unity implementation. Can you elaborate?

There is one thing I noticed that needs fixing on Android. The start rotation around the y axis is correct and the same for every orientation. For instance you open the app while having the phone in landscape left, you see the same as if you would open the app in any other orientation. However if you then rotate to portrait you will see a different rotation around the y axis. It has rotated -90 degrees. Rotate back to landscape left and it has rotated +90 degrees from the start position. On iOS it does function correctly. I am looking for a way to fix this.

maartenslaa commented:
I pushed the last fix for the incorrect rotation. The issue was the 'onCreate' method is called when the orientation changes, causing the headtracker to be reinitialised before the new screen orientation is detected by the ScreenOrientationHelper. Adding the android:configChanges:"orientation" to the manifest solves the issue because onCreate is not called when the orientation changes.

agalbachicar commented on sdk/screen_params.h:
```diff
@@ -28,6 +28,16 @@ void initializeAndroid(JavaVM* vm, jobject context);
                            float* out_width_meters, float* out_height_meters);
+
+// Aryzon multiple orientations
+enum ScreenOrientation {
```
I missed that we have the same enum definition in cardboard.h.

We should stick to have only one and use cardboard.h version instead. Please remove it from here.

agalbachicar commented on sdk/cardboard.cc:
```diff
@@ -119,6 +119,22 @@ void Cardboard_initializeAndroid(JavaVM* vm, jobject context) {
 #endif
 
+CardboardScreenOrientation CardboardScreenParameters_getScreenOrientation() {
+    switch (cardboard::screen_params::getScreenOrientation()) {
```
Implementation of this function should change to:

```C++
if (CARDBOARD_IS_NOT_INITIALIZED()) {
    return kUnknown;
}
return cardboard::screen_params::getScreenOrientation();
```

agalbachicar commented on sdk/include/cardboard.h:
```diff
@@ -126,6 +140,18 @@ void Cardboard_initializeAndroid(JavaVM* vm, jobject context);
+/// Returns the current device screen orientation
+///
+/// @return         CardboardScreenOrientation enum
+CardboardScreenOrientation CardboardScreenParameters_getScreenOrientation();
```
We should add a `@pre`requisite explaining that if the SDK is not initialized, it will always return `kUnknown`.

agalbachicar reviewed (changes requested):
Thanks for going over all my comments. There are some more, PTAL.

I will test the code in the app and Unity later this week, please bear with me a bit more. You are getting closer!

-----------------

There are a few pending discussions, let me expand:

My comment: 

> You mentioned to use another method instead of GetPose on the headtracker because this is missing the Unity interaction. I have tested with Unity and to my knowledge this did not impact the Unity implementation. Can you elaborate?

Your response:

> You mentioned to use another method instead of GetPose on the headtracker because this is missing the Unity interaction. I have tested with Unity and to my knowledge this did not impact the Unity implementation. Can you elaborate?

My reply to your comment:

I will test it on Unity, but my understanding is that it won't work out of the box if I only allowed other orientations. Will double check though.

Regardless of Unity, I should put a bit more pressure on my request to remove the head tracker changes. We have internal clients that are already managing the screen orientation and the API as is would incur in errors. I suggest two paths:

1- Create a new method that calls the current `GetPose()` where you implement the screen rotation logic.
2- Implement the logic outside the SDK, i.e. in the application.

Right now, I am biased towards 2 because it will require less SDK maintenance and testing.

@chaosemer @jballoffet any thoughts on this?

-----------

- [ ] Review thread safety once API is consolidated. --> @agalbachicar

-----------

May I ask you to remove all the occurrences of `Aryzon multiple orientations`?

maartenslaa commented:
Thanks for the comments! I've gone over them all and applied your recommendations.

Your comment:
> I will test it on Unity, but my understanding is that it won't work out of the box if I only allowed other orientations. Will double check though.

I've gone ahead and tested this, you are right this does not work as expected. The textures are not updated after rotating the device. However after adding one line this seems to be fixed: [#7834f08](https://github.com/googlevr/cardboard/pull/208/commits/7834f08921aebb592e2d8045ae9a30a1a52187cd)

Your comment:
> Regardless of Unity, I should put a bit more pressure on my request to remove the head tracker changes. We have internal clients that are already managing the screen orientation and the API as is would incur in errors. I suggest two paths:
> 
> 1- Create a new method that calls the current GetPose() where you implement the screen rotation logic.
> 2- Implement the logic outside the SDK, i.e. in the application.
> 
> Right now, I am biased towards 2 because it will require less SDK maintenance and testing.

I'm brainstorming here. The v1.4.1 head tracker returns an orientation based on landscape left, if you then would like to use landscape right you will need to compensate outside GetPose() or outside the SDK for not being landscape left. That sounds a little odd to me.

Otherwise the head tracker needs to not know about orientation, but that would mean stripping out the ekf_to_head_tracker_ and sensor_to_display. Applying these rotations outside the SDK doesn't really sound reasonable to me either.

If I had to choose option 1 sounds better to me, but why not use it as-is? How much of an impact will this be on maintenance?

agalbachicar commented:
@maartenslaa thanks a lot for your thoughts and interest in getting this through.

I've discussed the implementation details with @chaosemer , @jballoffet and @xinyunh0929 yesterday and we concluded the following:

- CardboardHeadTracker_getPose() should receive another argument that tells which is the base orientation, meaning landscape left, right, portrait or portrait upside down.
  - Why? Because it is hard to tell from the SDK point of view _whose_ orientation the user is interested in: phone, activity or a specific view.
  - Unity controls the rotation and, as you already pointed out, we should figure out how to properly track it.
- At the moment, we don't feel comfortable making the screen orientation tracker part of the API because of the above reasons. However, the code could be moved to the sample apps to show one possible implementation.

If you can wait a bit more, I'd like to dedicate a couple of hours to make this change part of a bigger change to the head tracker to also add the recentering functionality. For that, not only the implementation but also the API might change a bit. I'll share a small design document and we can take from there once we are settled.

I don't expect major changes here but yes some code refactor to move some pieces to the application instead of being behind the SDK API. 

What do you think?

BTW, a quick reply on your question about maintenance. We try to provide good QA support and adding this type of features increases considerably the combinatorial configurations of the applications to be tested. Even though some features are unrelated one to the other so there is no strong reason to have a full variety of apps that test every single possibility, it adds to the equation and increases the amount of work in QA rounds. We will do it because it's the right thing to do but also we try to seize the impact to end up with a good solution from most points of view.

maartenslaa commented:
Yes, I can see that will result in a cleaner and a bit more flexible SDK.

Will the ekf_to_head_tracker_ be set the first time GetPose is called? I think that will only work if the orientation on the first call to GetPose is the same as when you constructed the head tracker. Maybe that's something a recenter can help fix.

More thoughts on platform specific implementations:
Android
I'm not so happy with the current way to determine screen orientation on Android. That can change if the orientation is passed through from outside the SDK. For the HelloCardboard sample it might be best to check for screen size changes in the activity and do an orientation check once that happens. Seems more logical than checking every time the orientation changes by 1 degree.

iOS
A similar approach can be taken for iOS through the ViewController's viewWillTransitionTo() method. The way to determine the screen orientation can stay the same.

Unity
The Unity implementation already checks Screen.safeArea every frame so maybe add it there? [A check for DeviceOrientation changes is easy](https://docs.unity3d.com/ScriptReference/DeviceOrientation.html).


I'm learning a lot from this process so thanks! For future PR's I will contact one of you and discuss the best way to implement beforehand though.^^

agalbachicar commented:
> I'm learning a lot from this process so thanks! For future PR's I will contact one of you and discuss the best way to implement beforehand though.^^

Sure thing! We are getting great contributions so we are happy to discuss them first to ease the development process.

agalbachicar commented:
> I'll share a small design document and we can take from there once we are settled.

Design document was sent for internal review. We should getting a public link soon.

agalbachicar commented:
Design doc link: https://docs.google.com/document/d/1ijyL0wXe7XHn1p2YP_bW8Ziaacy7Eto9ZAloPbHokzs/edit?usp=sharing

Please feel free to comment your thoughts there. For this PR it only takes the multiple orientation support but feel free to consider adding your input to the recenter feature as well. Thanks for your patience.

jballoffet commented:
Hi @maartenslaa! I wanted to let you know that the multiple orientation support feature has been internally merged and it will be included in the next release. You'll find that the applied changes are the ones on #250 since it followed the approach described on the [design doc](https://docs.google.com/document/d/1ijyL0wXe7XHn1p2YP_bW8Ziaacy7Eto9ZAloPbHokzs/edit?usp=sharing) (which of course this PR couldn't follow since it was created before that document).

I want to thank you for sending this PR which allowed us to tackle this topic and to add it as a new feature to the Cardboard SDK. Again, thanks!

maartenslaa commented:
> Hi @maartenslaa! I wanted to let you know that the multiple orientation support feature has been internally merged and it will be included in the next release. You'll find that the applied changes are the ones on #250 since it followed the approach described on the [design doc](https://docs.google.com/document/d/1ijyL0wXe7XHn1p2YP_bW8Ziaacy7Eto9ZAloPbHokzs/edit?usp=sharing) (which of course this PR couldn't follow since it was created before that document).
> 
> I want to thank you for sending this PR which allowed us to tackle this topic and to add it as a new feature to the Cardboard SDK. Again, thanks!

Great to hear, thanks!

chaosemer commented:
PR #250 was merged in v1.7.0, which has the same purpose as this.
